ansible_base/jwt_consumer/common/auth.py

```diff
@@ -0,0 +1,203 @@
+            logger.debug(key)
+            self.log_and_raise("Returned key does not start and end with BEGIN/END PUBLIC KEY")
+        logger.info("Decryption key appears valid")
+        logger.debug(f"{key}")
+        return key
+
+    def map_user_fields(self, user, token):
+        user_needs_save = False
+        for attribute in self.mapped_user_fields:
+            old_value = getattr(user, attribute, None)
+            new_value = token.get(attribute, None)
+            if old_value != new_value:
+                logger.debug(f"Changing {attribute} for {user.username} from {old_value} to {new_value}")
+                setattr(user, attribute, new_value)
+                user_needs_save = True
+        if user_needs_save:
+            logger.info(f"Saving user {user.username}")
+            user.save()
+
+    def validate_token(self, token, decryption_key):
+        validated_body = None
+        required_fields = self.mapped_user_fields
+
+        # Ensure all of the internal pieces are part of the token but we don't want them in required_feidls or we will attempt to make them into the user object
+        local_required_field = ["sub", "claims", "exp"]
+        for field in required_fields:
+            if field not in local_required_field:
+                local_required_field.append(field)
+
+        # Decrypt the token
+        try:
+            logger.info("Decrypting token")
+            validated_body = jwt.decode(
+                token,
+                decryption_key,
+                audience="ansible-services",
+                options={"require": local_required_field},
+                issuer="ansible-issuer",
+                algorithms=["RS256"],
+            )
+        except jwt.exceptions.DecodeError as e:
+            self.log_and_raise(f"JWT decoding failed: {e}, check your key and generated token")
+        except jwt.exceptions.ExpiredSignatureError:
+            self.log_and_raise("JWT has expired")
+        except jwt.exceptions.InvalidAudienceError:
+            self.log_and_raise("JWT did not come for the correct audience")
+        except jwt.exceptions.InvalidIssuerError:
+            self.log_and_raise("JWT did not come from the correct issuer")
+        except jwt.exceptions.MissingRequiredClaimError as e:
+            self.log_and_raise(f"Failed to decrypt JWT: {e}")
+        except Exception as e:
+            self.log_and_raise(f"Unknown error occurred decrypting JWT ({e.__class__}) {e}")
+
+        logger.debug(validated_body)
+
+        return validated_body
+
+
+class JWTAuthentication(BaseAuthentication):
+    map_fields = default_mapped_user_fields
+
+    def authenticate(self, request):
+        common_auth = JWTCommonAuth(self.map_fields)
+        user, token = common_auth.parse_jwt_token(request)
+
+        if user:
+            self.process_user_data(user, token)
+            self.process_permissions(user, token.get("claims", None), token)
+
+            return user, None
```

Comment:
awx oauth authentication class returns the oauth token as the second tuple value. I wonder if we should return the token here?

Reply:
It looks like the only one from the rest framework that returns the token is the token auth https://github.com/encode/django-rest-framework/blob/master/rest_framework/authentication.py. We have a JWT token but I don't think it would be super useful to return that since they are pretty short lived.